### PR TITLE
[Bulls & cows]: Added tests, improved description

### PR DESCRIPTION
**Pull request by sergii-nosachenko**
## Description

Added unit tests, and improved task description.

### Screenshots

Screenshot 2024-01-17 at 15 55 52
Screenshot 2024-01-17 at 15 55 20

### CODE REVIEW

**Review by alexandra-protyanova — APPROVED** [review]

File: src/modules/calculateRandomNumber.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+module.exports = {
+  /**
+   * Calculate a random 4-digit number that does not start with 0
+   * and does not contain any duplicate digits.
+   *
+   * @return {number} A random 4-digit number
+   */
+  calculateRandomNumber: () => { /* Write your code here */ },
+};
```
Comment (mgrinko) [inline src/modules/calculateRandomNumber.js:11]: Usualy we create a function and then import it

```suggestion
/**
 * Calculate a random 4-digit number that does not start with 0
 * and does not contain any duplicate digits.
 *
 * @return {number}
 */
function calculateRandomNumber() => {
  /* Write your code here */
}

module.exports = { calculateRandomNumber };
```

File: src/modules/calculateRandomNumber.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+'use strict';
+
+module.exports = {
+  /**
+   * Calculate a random 4-digit number that does not start with 0
+   * and does not contain any duplicate digits.
+   *
+   * @return {number} A random 4-digit number
+   */
+  calculateRandomNumber: () => { /* Write your code here */ },
```
Comment (mgrinko) [inline src/modules/calculateRandomNumber.js:10]: Why `calculate`?
We `generate` it.

File: tests/calculateRandomNumber.test.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+    expect(result).toBeGreaterThan(999);
+    expect(result).toBeLessThan(10000);
+  });
+
+  test('should not start with 0', () => {
+    for (let i = 0; i < 100; i++) {
+      const result = calculateRandomNumber();
+
+      expect(result.toString()[0]).not.toBe('0');
```
Comment (mgrinko) [inline tests/calculateRandomNumber.test.js:25]: if `result` is a number
then just check if it > 999

File: tests/calculateRandomNumber.test.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    }
+  });
+
+  test('should not contain duplicate digits', () => {
+    for (let i = 0; i < 100; i++) {
+      const result = calculateRandomNumber().toString();
+      const digits = new Set(result);
```
Comment (mgrinko) [inline tests/calculateRandomNumber.test.js:32]:
```suggestion
      const result = calculateRandomNumber();
      const digits = new Set(result.toString());
```

File: tests/calculateRandomNumber.test.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+      }
+    }
+
+    const averageCount = iterations * 4 / 10;
+
+    for (const count of digitCounts) {
+      const deviation = Math.abs(count - averageCount) / averageCount;
+
+      expect(deviation).toBeLessThanOrEqual(tolerance);
+    }
```
Comment (mgrinko) [inline tests/calculateRandomNumber.test.js:57]: This logic is too complex
I'd just check that we have at least 50 different values
This is a Node.js course but not Randomization Algo :)

File: tests/ensureValidUserInput.test.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    const validInputs = [
+      '1234',
+      '5678',
+      '9012',
+      '3456',
+      '7890',
+    ];
+
+    for (const validInput of validInputs) {
+      expect(ensureValidUserInput(validInput)).toBe(Number(validInput));
```
Comment (mgrinko) [inline tests/ensureValidUserInput.test.js:59]: Why does this function returns `null` or `number`?
It is better to have `isValid` function that just checks input validity

File: tests/getBullsAndCows.test.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+  test('returns correct number of bulls and cows', () => {
+    expect(getBullsAndCows(1234, 1243))
+      .toEqual({
+        bulls: 2,
+        cows: 2,
+      });
+  });
+
+  test('handles cases with repeating digits correctly', () => {
+    expect(getBullsAndCows(1223, 3211))
```
Comment (mgrinko) [inline tests/getBullsAndCows.test.js:47]: The function should newer receive numbers with repeating digits!!

**Review by mgrinko — CHANGES_REQUESTED** [review]

**Pull request merged**